core/lib/multivm/src/versions/vm_fast/validation_tracer.rs

```diff
@@ -0,0 +1,215 @@
+/// the spam comes from multiple addresses, that doesn't work.
+///
+/// We want to ensure that a spammer has to pay for every account that fails validation. This is
+/// achieved by limiting what the code of a custom account is allowed to do. If we allowed access
+/// to things like time, a validation that fails in the sequencer could be crafted for free, so we
+/// don't.
+///
+/// However, we want to give access to storage. A spammer has to pay for changing storage but
+/// could change just one storage slot to invalidate transactions from many accounts. To prevent
+/// that, we make sure that the storage slots accessed by different accounts are disjoint by only
+/// allowing access to storage in the account itself and slots derived from the account's address.
+///
+/// Our rules are an extension of the rules are outlined in EIP-7562.
+///
+/// This tracer enforces the rules by checking what the code does at runtime, even though the
+/// properties checked are supposed to always hold for a well-written custom account. Proving
+/// that a contract adheres to the rules ahead of time would be challenging or even impossible,
+/// considering that contracts that the code depends on may get upgraded.
+#[derive(Debug, Default)]
+pub struct ValidationTracer {
+    in_validation: bool,
+    add_return_value_to_allowed_slots: bool,
+
+    slots_obtained_via_keccak: HashSet<U256>,
+    trusted_addresses: HashSet<Address>,
+
+    user_address: Address,
+    trusted_storage: HashSet<(Address, U256)>,
+    /// These location's values are added to [Self::trusted_addresses] to support upgradeable proxies.
+    storage_containing_trusted_addresses: HashSet<(Address, U256)>,
+
+    validation_error: Option<ViolatedValidationRule>,
+}
+
+impl Tracer for ValidationTracer {
+    fn before_instruction<OP: OpcodeType, S: GlobalStateInterface>(&mut self, state: &mut S) {
+        if !self.in_validation {
+            return;
+        }
+
+        match OP::VALUE {
+            // Out of gas once means out of gas for the whole validation, as the EIP forbids handling out of gas errors
+            Ret(Panic) if state.current_frame().gas() == 0 => {
+                self.set_error(ViolatedValidationRule::TookTooManyComputationalGas(0))
+            }
+
+            ContextMeta => self.set_error(ViolatedValidationRule::TouchedDisallowedContext),
+
+            StorageRead => {
+                let address = state.current_frame().address();
+                let caller = state.current_frame().caller();
+
+                // Can unwrap because the instruction pointer does not point to a panic instruction
+                let pc = state.current_frame().program_counter().unwrap();
+                let word = pc / 4;
+                let part = pc % 4;
+                let instruction =
+                    state.current_frame().read_contract_code(word).0[3 - part as usize];
+                let slot = state.read_register((instruction >> 16) as u8 & 0b1111).0;
```

Comment:
It is extremely low-level. I didn't want to design an interface for getting instruction targets because this is the only place it is needed, so I don't know what a good interface would be like.